MM-28649 Add Mobile E2E documentation (#682)
mattermost/mattermost-developer-documentation: site/content/contribute/mobile/e2e/ios.md

---
title: "Setup and Run iOS E2E"
date: 2020-09-01T09:00:00-00:00
weight: 4
subsection: Mobile End-to-End Testing
---

#### Local setup
1. Install [applesimutils](https://github.com/wix/AppleSimulatorUtils)
```
brew tap wix/brew
brew install applesimutils
```

> 2. Set XCode's build location so that the built app, especially debug, is expected at project's location instead at the Library's folder which is unique/hashed.

Reviewer: ```suggestion
2. Set XCode's build location so that the built app, especially debug, is expected at the project's location instead of the Library's folder which is unique/hashed.
```

> - Open XCode, then go to "XCode > Preferences > Locations"

Reviewer: ```suggestion
- Open XCode, then go to **XCode > Preferences > Locations**.
```

> - Open XCode, then go to **XCode > Preferences > Locations**.

Reviewer: ```suggestion
1. Open XCode, then go to **XCode > Preferences > Locations**.
```

> - Under "Derived Data", click "Advanced..."

Reviewer: ```suggestion
- Under **Derived Data**, click **Advanced...**.
```

> - Select "Custom" and set to "Relative to Workspace", then set Products as "Build/Products".

Reviewer: ```suggestion
- Select **Custom > Relative to Workspace**, then set **Products** as **Build/Products**.
```

> - Select **Custom > Relative to Workspace**, then set **Products** as **Build/Products**.

Reviewer: ```suggestion
3. Select **Custom > Relative to Workspace**, then set **Products** as **Build/Products**.
```

> - Click "Done" to take effect the changes.

Reviewer 2: ```suggestion
- Click "Done" to save the changes.
```

Reviewer: Suggest "Done" to **Done**.

> - Click "Done" to save the changes.

Author: ```suggestion
- Click **Done** to save the changes.
```

> - Click **Done** to save the changes.

Reviewer: ```suggestion
4. Click **Done** to save the changes.
```

> #### Test run on debug mode

Reviewer 2: ```suggestion
#### Test run in debug mode
```

> This is the typical flow for local development and test writing.

Reviewer: ```suggestion
This is the typical flow for local development and test writing:
```

> 1. On first terminal, `make run` or `make run-ios` at the root folder.

Reviewer: ```suggestion
1. In one terminal window, run `make run` or `make run-ios` from within the root folder.
```

> 1. In one terminal window, run `make run` or `make run-ios` from within the root folder.

Reviewer: ```suggestion
1. In one terminal window, run `make run` or `make run-ios` from the root folder.
```

> 2. On second terminal, `npm i` then `npm run e2e:ios-test` at `/detox` folder.

Reviewer: ```suggestion
2. In another terminal window, run `npm i` then `npm run e2e:ios-test` from within the `/detox` folder.
```

> #### Test run on release mode

Reviewer 2: ```suggestion
#### Test run in release mode
```

> 2. In another terminal window, run `npm i` then `npm run e2e:ios-test` from within the `/detox` folder.

Reviewer: ```suggestion
2. In another terminal window, run `npm i` then `npm run e2e:ios-test` from the `/detox` folder.
```

This is the typical flow for CI test run.

> 1. Build release app by `make ios-sim-x86_64` at the root folder or `npm run e2e:ios-build-release` at the `/detox` folder.

Reviewer: ```suggestion
1. Build the release app by running `make ios-sim-x86_64` from within the root folder or `npm run e2e:ios-build-release` from within the `/detox` folder.
```

> 2. Run test by `npm run e2e:ios-test-release` at `/detox` folder.

Reviewer: ```suggestion
2. Run the test using `npm run e2e:ios-test-release` from within the `/detox` folder.
```

> 1. Build the release app by running `make ios-sim-x86_64` from within the root folder or `npm run e2e:ios-build-release` from within the `/detox` folder.

Reviewer: ```suggestion
1. Build the release app by running `make ios-sim-x86_64` from the root folder or `npm run e2e:ios-build-release` from within the `/detox` folder.
```

---
title: "Setup and Run iOS E2E"
date: 2020-09-01T09:00:00-00:00
weight: 4
subsection: Mobile End-to-End Testing
---

#### Local setup

1. Install [applesimutils](https://github.com/wix/AppleSimulatorUtils):
```
brew tap wix/brew

> brew install applesimutils

Reviewer 3: is this required? can't we use the Xcode simulator commands for this? or does detox use this under the hood ?

Reviewer 4: Detox is using this to control iOS simulator - start/stop/reload/install/uninstall/set permission etc.

```
2. Set XCode's build location so that the built app, especially debug, is expected at the project's location instead of the Library's folder which is unique/hashed.
3. Open XCode, then go to **XCode > Preferences > Locations**.
4. Under **Derived Data**, click **Advanced...**.
5. Select **Custom > Relative to Workspace**, then set **Products** as **Build/Products**.

> 6. Click **Done** to save the changes.

Reviewer 3: Isn't this already defined by the iOS project? can we verify this is actually required? specially  as there is no way to do this on the CI

Reviewer 4: Yes, this is required in local development so we know where to pick up the [binary](https://github.com/mattermost/mattermost-mobile/blob/master/detox/.detoxrc.json#L6). If not, the binary will be placed in the default XCode folder with non-deterministic and dynamically generated hash folder. No change in CI - still on [`/build-ios` folder](https://github.com/mattermost/mattermost-mobile/blob/master/fastlane/Fastfile#L362-L365).

#### Test run in debug mode

This is the typical flow for local development and test writing:

> 1. In one terminal window, run `make run` or `make run-ios` from the root folder.

Reviewer 3: same as with Android.

Reviewer 4: In iOS debug, we're building thru `react-native` not via [detox](https://github.com/mattermost/mattermost-mobile/blob/master/detox/.detoxrc.json#L5-L11). So either of these make commands are accurate and correct.

Reviewer 4: No change yet as there's no equivalent `npm` script at the moment. Will update in the future once available.

Reviewer 3: I'll make the changes to this doc later on. but it could've been replaced with `npx react-native run-ios` in the meantime.
